sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotUseLiteralBoolInAssertions.cs

```diff
@@ -0,0 +1,120 @@
+
+        private static Dictionary<KnownType, HashSet<string>> trackedTypeAndMethods =
+            new Dictionary<KnownType, HashSet<string>>
+            {
+                [KnownType.Xunit_Assert] = new HashSet<string>
+                {
+                    "Equal", "False", "NotEqual", "Same", "StrictEqual", "NotSame"
+                },
+
+                [KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_Assert] = new HashSet<string>
+                {
+                    "AreEqual", "AreNotEqual", "AreSame", "IsFalse", "IsTrue"
+                },
+
+                [KnownType.NUnit_Framework_Assert] = new HashSet<string>
+                {
+                    "AreEqual", "AreNotEqual", "AreNotSame", "AreSame", "False",
+                    "IsFalse", "IsTrue", "That", "True"
+                },
+
+                [KnownType.System_Diagnostics_Debug] = new HashSet<string>
+                {
+                    "Assert"
+                }
+            };
+
+        private static List<SyntaxKind> boolLiterals =
+            new List<SyntaxKind>
+            {
+                SyntaxKind.TrueLiteralExpression,
+                SyntaxKind.FalseLiteralExpression
+            };
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var invocation = (InvocationExpressionSyntax)c.Node;
+
+                    var symbolInfo = c.SemanticModel.GetSymbolInfo(invocation);
+                    var methodSymbol = symbolInfo.Symbol ?? symbolInfo.CandidateSymbols.FirstOrDefault();
+
+                    if (methodSymbol != null &&
+                        invocation.ArgumentList != null &&
+                        IsTrackedMethod(methodSymbol) &&
+                        IsFirstOrSecondArgumentABoolLiteral(invocation.ArgumentList.Arguments))
+                    {
+                        c.ReportDiagnostic(Diagnostic.Create(rule, invocation.GetLocation()));
+                    }
+                },
+                SyntaxKind.InvocationExpression);
+        }
+
+        private static bool IsFirstOrSecondArgumentABoolLiteral(
+            SeparatedSyntaxList<ArgumentSyntax> arguments)
+        {
+            switch (arguments.Count)
+            {
+                case 0:  return false;
+                case 1:  return IsBooleanLiteral(arguments[0]);
+                default: return IsBooleanLiteral(arguments[0]) || IsBooleanLiteral(arguments[1]);
+            }
+        }
+
+        private static bool IsBooleanLiteral(ArgumentSyntax argument)
+        {
+            return argument.Expression.IsAnyKind(boolLiterals);
+        }
+
+        private static bool IsTrackedMethod(ISymbol methodSymbol)
+        {
+            return trackedTypeAndMethods
+                .Where(kvp => methodSymbol.ContainingType.Is(kvp.Key))
+                .Any(kvp => kvp.Value.Contains(methodSymbol.Name));
```

Comment:
Probably a bit less pretty, but should be more performant:
```
ISet<string> methods;
return trackedTypeAndMethods.TryGetValue(methodSymbol.ContainingType.Name, out methods) &&
    methods.Contains(methodSymbol.Name);
```

Reply:
I like that version, but it won't work. `TypeHelper.Is` calls `TypeHelper.IsMatch` which does slightly bit more than string comparison.